src/functions/authoriser/framework/handler.ts

```diff
@@ -1,6 +1,49 @@
-import { CustomAuthorizerEvent, CustomAuthorizerResult, Context } from 'aws-lambda';
+import * as util from 'util';
+import { CustomAuthorizerEvent, CustomAuthorizerResult } from 'aws-lambda';
+import * as jwksClient from 'jwks-rsa';
+import nodeFetch from 'node-fetch';
+import { AdJwtVerifier } from '../application/AdJwtVerifier';
 
-const newCustomAuthorizerResult = (principalId: string, effect: string, resource: string)
+const ensureNotNullOrEmpty = (val: any, fieldName: string) => {
+  if (val === null || val === undefined || val === '') {
+    throw `Error: ${fieldName} is null or empty`;
+  }
+};
+
+/**
+ * Creates an AdJwtVerifier that can verify the authenticity of Azure Active Directory JWTs.
+ */
+const createAdJwtVerifier = async (): Promise<AdJwtVerifier> => {
+  const tenantId: string = process.env.DVSA_MES_AzureAD_TenantId || '';
+  const applicationId: string = process.env.DVSA_MES_AzureAD_ClientId || '';
+
+  ensureNotNullOrEmpty(tenantId, 'tenantId');
+  ensureNotNullOrEmpty(applicationId, 'applicationId');
+
+  const openidConfig = await (await nodeFetch(
+    `https://login.microsoftonline.com/${tenantId}/.well-known/openid-configuration`,
+  )).json();
+
+  if (openidConfig.error_description) {
+    throw `Failed to get openid configuration: ${openidConfig.error_description}`;
+  }
+
+  const jwksclient = jwksClient({
+    jwksUri: openidConfig.jwks_uri,
+    cache: true,
```

Comment:
do we know how this is being cached? given this is running in lambda, feels like magic to me

Reply:
Its just being cached in memory, so the caching here only for that specific instance of the authoriser.  Yes, multiple instances of the authoriser will make their own HTTP requests to get the public keys.

Do you think we should implement caching in a store that is shared by all instances?  Its a lot more work.  I just thought that caching within each instance limits the requests enough.

Reply:
ok, should be good enough. This would be a good candidate for something to log

Reply:
Capturing offline discussion/conclusion.  The goal is to be able to monitor how often web requests are being made to get new keys.  The `jwksClient` package doesn't seem to have an explicit way of doing this, but it does use the `debug` package, so if we enable debugging for the library by setting the `DEBUG` envvar to include the value `jwks`, then we get plenty of debug about what it is doing internally.  Its too much debug info to leave enabled for production though.

The current thinking is to leave this for now, to focus on getting an initial version of the custom authoriser complete.  Then we can look at how to do this monitoring.